desc/magnetic_fields.py

```diff
@@ -1637,6 +1644,351 @@ def from_surface(
+        data : dict of ndarray
+            Data computed so far, generally output from other compute functions
+        override_grid : bool
+            If True, override the user supplied grid if necessary and use a full
+            resolution grid to compute quantities and then downsample to user requested
+            grid. If False, uses only the user specified grid, which may lead to
+            inaccurate values for surface or volume averages.
+
+        Returns
+        -------
+        data : dict of ndarray
+            Computed quantity and intermediate variables.
+
+        """
+        if isinstance(names, str):
+            names = [names]
+        if grid is None:
+            grid = LinearGrid(theta=2 * self.M_well, zeta=8, NFP=self.NFP, sym=False)
+        elif not isinstance(grid, _Grid):
+            raise TypeError(
+                "must pass in a Grid object for argument grid!"
+                f" instead got type {type(grid)}"
+            )
+
+        if params is None:
+            params = get_params(names, obj=self)
+        if transforms is None:
+            transforms = get_transforms(names, obj=self, grid=grid, **kwargs)
+        if data is None:
+            data = {}
+        profiles = {}
+
+        data = compute_fun(
+            self,
+            names,
+            params=params,
+            transforms=transforms,
+            profiles=profiles,
+            data=data,
+            **kwargs,
+        )
+        return data
+
+    @property
+    def NFP(self):
+        """int: Number of (toroidal) field periods."""
+        return self._NFP
+
+    @NFP.setter
+    def NFP(self, NFP):
+        assert isposint(NFP), f"NFP should be a positive integer, got {type(NFP)}"
+        self.change_resolution(NFP=NFP)
+
+    @property
+    def L_well(self):
+        """int: Radial resolution of the magnetic well parameters well_l."""
+        return self._L_well
+
+    @property
+    def M_well(self):
+        """int: Number of spline points in the magnetic well parameters well_l."""
+        return self._M_well
+
+    @property
+    def L_shift(self):
+        """int: Radial resolution of x_lmn."""
+        return self._L_shift
+
+    @property
+    def M_shift(self):
+        """int: Poloidal resolution of x_lmn."""
+        return self._M_shift
+
+    @property
+    def N_shift(self):
+        """int: Toroidal resolution of x_lmn."""
+        return self._N_shift
+
+    @property
+    def well_basis(self):
+        """ChebyshevPolynomial: Spectral basis for B_lm."""
+        return self._well_basis
+
+    @property
+    def shift_basis(self):
+        """FourierZernikeBasis: Spectral basis for x_lmn."""
+        return self._shift_basis
+
+    @optimizable_parameter
+    @property
+    def B_lm(self):
+        """ndarray: Omnigenity magnetic well shape parameters."""
+        return self._B_lm
+
+    @B_lm.setter
+    def B_lm(self, B_lm):
+        self._B_lm[:] = B_lm
```

Comment:
jax arrays aren't mutable so this should be just a regular assignment
`self._B_lm = B_lm`